### PR TITLE
Scale_layer: enable scale layer to handle inputs in batch

### LINKED ISSUE #13626: dnn: assertion failure with multiple images
- OpenCV => 4.0.1-dev
- Operating System / Platform => Windows 64 Bit
- Compiler => mingw64 7.2.0
##### Detailed description
using the[ face_detector_uint8.pb](https://github.com/opencv/opencv_3rdparty/raw/8033c2bc31b3256f0d461c919ecc01c2428ca03b/opencv_face_detector_uint8.pb) model with multiple images (using dnn::blobFromImages), i get :
```
OpenCV(4.0.1-dev) Error: Assertion failed (start <= (int)shape.size() && end <= (int)shape.size() && start <= end) in total, file C:/p/opencv/modules/dnn/include/opencv
2/dnn/shape_utils.hpp, line 161
```

debugging it, i found, that in the 2nd ScaleLayer, the `endAxis` exceeds the size of the `inpShape`:

```
Thread 1 hit Breakpoint 9, cv::dnn::ScaleLayerImpl::forward (this=0xa5c1090, inputs_arr=..., outputs_arr=..., internals_arr=...)
    at C:\p\opencv\modules\dnn\src\layers\scale_layer.cpp:94
94              CV_Assert(total(inpShape, axis, endAxis) == numWeights);
(gdb) p inpShape
$16 = std::vector of length 4, capacity 4 = {3, 32, 24, 32}
(gdb) p axis
$17 = 0
(gdb) p endAxis
$18 = 5 // this is 2 with one image, and 5 with 3 of them !
```
thus the assert fails here: 

https://github.com/opencv/opencv/blob/7e2ebecd52232d02e2d53f3e073c076be73e3470/modules/dnn/include/opencv2/dnn/shape_utils.hpp#L160-L161

##### Steps to reproduce
    ```
    dnn::Net net = dnn::readNetFromTensorflow("opencv_face_detector_uint8.pb","opencv_face_detector.pbtxt");
    Mat f1=imread("../img/h1.png");
    Mat f2=imread("../img/h2.png");
    Mat f3=imread("../img/h3.png");
    vector<Mat> vf {f1, f2, f3};
    Mat blob = dnn::blobFromImages(vf, 1, Size(128,96), Scalar(104, 177, 123, 0), false, false);
    net.setInput(blob);
    Mat res = net.forward("detection_out");
    ```

### LINKED ISSUE #16630: dnn(Scale): errornous handling of batch inputs in untrainable scale mode
##### System information (version)
- OpenCV => 3.4

##### Detailed description

The untrainable mode in ScaleLayer uses the second input as scale factors for the first input.
```
outputs[0] = inputs[0] * inputs[1] // elementwise multiplication with broadcasting
```

The code which performs the above operation appears to assume that the batch size is one (but there is no assertion to enforce the requirement).

https://github.com/opencv/opencv/blob/3efa78311a837b7a06d81f2f359ca98e6d26609a/modules/dnn/src/layers/scale_layer.cpp#L86-L94

`weights` is reshaped (assuming that batch size is one?).

https://github.com/opencv/opencv/blob/3efa78311a837b7a06d81f2f359ca98e6d26609a/modules/dnn/src/layers/scale_layer.cpp#L98-L104

This code attempts to look for broadcasting options.  It does so by attempting to match `numWeights` (which is `weights.total()`) to a subrange in the input tensor starting from `axis`. This logic will fail to find a match and depending on input shapes will either lead to incorrect outputs or tip-off downstream assertions 

**Example:**
```
input shape: 2 3 3 3
weights input: 2 3 1 1
```

The shapes are compatible for an elementwise operation. The current logic will fail to see it.

`weights` is reshaped to `6 1`. The matching code tries to look for a subrange in the input `(2 3 3 3)` which totals to 6 but there is no such range.

This particular example triggers an assertion.

```
(-215:Assertion failed) start <= (int)shape.size() && end <= (int)shape.size() && start <= end in function 'total'
```

##### Steps to reproduce
A test for PR16436 (not merged yet) can reproduce it:

https://github.com/opencv/opencv/blob/a42268678eac90fb7d5f09d16e3a2db7167d2641/modules/dnn/test/test_darknet_importer.cpp#L547-L548

Change `testDarknetLayer("scale_channels", false, false);` to `testDarknetLayer("scale_channels", false, true);`

### LINKED ISSUE #17063: OpenCV DNN - Assertion Failed on Batch
##### System information (version)
- OpenCV => 4.3.0-dev
- Operating System / Platform => Ubuntu 18.04
- Compiler => Python 3.6.8

##### Detailed description

Using OpenCV DNN with an EfficientNetB0 saved using Tensorflow I can input a batch of 1 image but it breaks while doing the forward when I input a batch higher than that. The error is:

> Traceback (most recent call last):
>   File "efficientnet_batch_error.py", line 18, in <module>
>     output = model.forward(output_layers)
> cv2.error: OpenCV(4.3.0-dev) /home/computer/workspace/opencv/modules/dnn/include/opencv2/dnn/shape_utils.hpp:171: error: (-215:Assertion failed) start <= (int)shape.size() && end <= (int)shape.size() && start <= end in function 'total'

##### Steps to reproduce

```
from cv2 import dnn
import numpy as np


model_bytes = open('b0.pb', 'rb').read()
model = dnn.readNetFromTensorflow(model_bytes)
del model_bytes
ln = model.getLayerNames()
output_layers = [ln[i[0] - 1] for i in model.getUnconnectedOutLayers()]
output_layers = output_layers[-1]

# This works
input_images = np.random.random((1, 3, 224, 224))
model.setInput(input_images)
output = model.forward(output_layers)
print(output.shape)

# This breaks
input_images = np.random.random((32, 3, 224, 224))
model.setInput(input_images)
output = model.forward(output_layers)
print(output.shape)
```

The model can be found [here](https://drive.google.com/open?id=1xnV2uydgaUffVTCwBhz51qN2qcZKunVa).

### PR DESCRIPTION
**Pull request by ashishkrshrivastava**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/729

resolves #16630 
resolves #17063
resolves #13626

```
opencv_extra=scale_batchsize
```

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### This PR Modifies
Handling of inputs in batch for scale layer. 

Changes are done considering two cases:
1 - Weights and Biases are constants, in this case dimension could be either [1, C, H, W] or [C, H, W].
2 - Weights and Biases are non-constants ( say input_1 ), i.e. they are calculated at runtime. so their shapes will be [N, C, H, W] where N will be same as batch size of input_0.

So Basically changes are added for Point 2. i.e. untrainable parameters/ scales with batch size greater than 1. Before scale layer was only supporting Point 1. i.e. trainable weights, ignoring the fact that scales can have batchsize > 1.

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.3.0:16.04
build_image:Custom Win=openvino-2020.1.0
build_image:Custom Mac=openvino-2020.1.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
test_opencl:Custom=ON
test_bigdata:Custom=1
test_filter:Custom=*Layer_Test*:*cale*
```

### CODE REVIEW

**Comment by YashasSamaga** [comment]
There is a `scale_channels` test which was added recently but batch testing was disabled due to this bug.

https://github.com/opencv/opencv/blob/3a99ebb446cec957fcbbb221933e0cd2a568dd8e/modules/dnn/test/test_darknet_importer.cpp#L557-L558

You have to change the last argument to `true` to enable the test on batches.

**Comment by ashishkrshrivastava** [comment]
>  You have to change the last argument to true to enable the test on batches.

@YashasSamaga thanks, Done !

File: modules/dnn/test/test_layers.cpp [hunk 1]
```diff
@@ -690,22 +697,27 @@ TEST_P(Scale_untrainable, Accuracy)
...
 
 INSTANTIATE_TEST_CASE_P(Layer_Test, Scale_untrainable, Combine(
 /*input size*/   Values(Vec4i(2, 3, 4, 5)),
 /*axis, #dims*/  Values(Vec2i(0, 1), Vec2i(0, 2), Vec2i(0, 3), Vec2i(0, 4),
-                                     Vec2i(1, 1), Vec2i(1, 2), Vec2i(1, 3),
-                                                  Vec2i(2, 1), Vec2i(2, 2),
-                                                               Vec2i(3, 1)),
+                                     Vec2i(1, 2), Vec2i(1, 3), Vec2i(1, 4),
+                                     Vec2i(2, 2), Vec2i(2, 3),
+                                     Vec2i(3, 2)),
```
[the remark this answers is not in the document]
Comment (ashishkrshrivastava) [reply modules/dnn/test/test_layers.cpp]: Sorry, couldn't understand.

> And just add one more parameter for batch size.

by this you mean `axis, dims, batchsize` ? 
Batchsize is always 2 in these tests and original parameters are still there. `axis, dims`.
I just changed dims by adding 1 because previously batchsize wasn't considered in dims.

File: modules/dnn/src/layers/scale_layer.cpp [hunk 2]
```diff
@@ -78,64 +78,94 @@ class ScaleLayerImpl CV_FINAL : public ScaleLayer
...
-        Mat weights = hasWeights ? (blobs.empty() ? inputs[1] : blobs[0]).reshape(1, 1) : Mat();;
-        Mat bias = hasBias ? (blobs.empty() ? inputs[1] : blobs.back()).reshape(1, 1) : Mat();
-
         MatShape inpShape = shape(inpBlob);
+        const int numSamples = inpShape[0];
+
+        const int numWeightsSamples = (hasWeights || hasBias) && blobs.empty() ? numSamples : 1;
+
+        Mat weights = hasWeights ? (blobs.empty() ? inputs[1] : blobs[0]).reshape(1 , 1) : Mat();
+        Mat bias = hasBias ? (blobs.empty() ? inputs[1] : blobs.back()).reshape(1 , 1) : Mat();
```
[the remark this answers is not in the document]
Comment (ashishkrshrivastava) [reply modules/dnn/src/layers/scale_layer.cpp]: No, we dont need really. just wanted to initialize scalar variable above Mat variables. I will change that.

File: modules/dnn/test/test_layers.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dkurt) [reply modules/dnn/test/test_layers.cpp]: So why do we need to modify this test? Can you please clarify a bit

File: modules/dnn/src/layers/scale_layer.cpp [hunk 3]
```diff
@@ -29,7 +29,7 @@ class ScaleLayerImpl CV_FINAL : public ScaleLayer
     {
         setParamsFrom(params);
         hasBias = params.get<bool>("bias_term", false);
-        axis = params.get<int>("axis", 1);
+        axis = (params.get<int>("axis", 1) > 0) ? params.get<int>("axis", 1) : 1;
```
Comment (YashasSamaga) [inline modules/dnn/src/layers/scale_layer.cpp]: Is this correct? Why can't axis be negative?

File: modules/dnn/test/test_layers.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ashishkrshrivastava) [reply modules/dnn/test/test_layers.cpp]:
```cpp
/*axis, #dims*/  Values(Vec2i(0, 1), Vec2i(0, 2), Vec2i(0, 3), Vec2i(0, 4),
                                     Vec2i(1, 1), Vec2i(1, 2), Vec2i(1, 3),
                                                  Vec2i(2, 1), Vec2i(2, 2),
                                                               Vec2i(3, 1)),
```

shape of input [ N, C, H, W ] 

`Vec2i(0, 1) - axis = 0, dim = 1 [ N ]` fine
`Vec2i(0, 2) - axis = 0, dim = 2 [ N, C ]` fine
`Vec2i(0, 3) - axis = 0, dim = 2 [ N, C, H ]` fine
`Vec2i(0, 4) - axis = 0, dim = 2 [ N, C, H, W ]` fine
`Vec2i(1, 1) - axis = 1, dim = 1 [ N ]` , seems incorrect 
`Vec2i(1, 2) - axis = 1, dim = 2 [ N, C ]` - fine
`Vec2i(1, 3) - axis = 1, dim = 3 [ N, C, H ]` - fine
`Vec2i(2, 1) - axis = 2, dim = 1 [ N ]`, seems incorrect 
`Vec2i(2, 2) - axis = 2, dim = 2 [ N, H ]`- fine, can be broadcasted as  [ N, 1, H ] 
`Vec2i(3, 1) - axis = 3, dim = 1 [ N ]`- seems incorrect

So I removed these three tests [ (1, 1), (2, 1), (3, 1)]. and added three more test that are possible [ (1, 4), (2, 3), (3, 2) ]:
`Vec2i(1, 4) - axis = 1, dim = 4 [ N, C, H, W ]`- fine
`Vec2i(2, 3) - axis = 2, dim = 3 [ N, C, H ]`- fine
`Vec2i(3, 2) - axis = 3, dim = 2 [ N, W ]`- fine, can be broadcasted as [ N, 1, 1, W ]

```cpp

/*axis, #dims*/  Values(Vec2i(0, 1), Vec2i(0, 2), Vec2i(0, 3), Vec2i(0, 4),
                                     Vec2i(1, 2), Vec2i(1, 3), Vec2i(1, 4),
                                     Vec2i(2, 2), Vec2i(2, 3),
                                     Vec2i(3, 2)),
```

File: modules/dnn/src/layers/scale_layer.cpp (hunk 3 above)
Comment (ashishkrshrivastava) [reply modules/dnn/src/layers/scale_layer.cpp]: Good point, Thanks.

Comment (ashishkrshrivastava) [reply modules/dnn/src/layers/scale_layer.cpp]: @YashasSamaga, please have a look now.

Comment (YashasSamaga) [reply modules/dnn/src/layers/scale_layer.cpp]: I think you're looking for this:

https://github.com/opencv/opencv/blob/ca23c0e630ba9c4fc147c4913d6238f5de3bbcd9/modules/dnn/src/layers/fully_connected_layer.cpp#L118

Comment (ashishkrshrivastava) [reply modules/dnn/src/layers/scale_layer.cpp]: @YashasSamaga not exactly, but yeah

```
axis = clamp(axis , inpBlob.dims) ? clamp(axis , inpBlob.dims) : 1;
CV_Assert(axis > 0 && axis < inpBlob.dims);
```
this would work. 

actually here in this loop, `numWeightsPerSample` is total weights excluding batchsize, so we need to check with axis >= 1.
```
for (endAxis = axis + 1; endAxis <= inpBlob.dims; ++endAxis)
        {
            if (total(inpShape, axis, endAxis) == numWeightsPerSample)
                break;
        }
endAxis = (numWeightsPerSample == 1) ? 1 : endAxis;
```

File: modules/dnn/test/test_layers.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dkurt) [reply modules/dnn/test/test_layers.cpp]: Can you please add a separate test only for new functionality? If I'm not mistaken, we try so solve batching in Scale Layer. So while
```
1 x C x H x W
1 x C
```
work can we test
```
N x H x W x C
1 x C
```
and
```
N x C x H x W
N x C
```
?

[the remark this answers is not in the document]
Comment (ashishkrshrivastava) [reply modules/dnn/test/test_layers.cpp]: sure. they work. Let me add tests for them.

[the remark this answers is not in the document]
Comment (ashishkrshrivastava) [reply modules/dnn/test/test_layers.cpp]: https://github.com/opencv/opencv/blob/2842116ebd717356ecc8f222284161ae5f0cb4a9/modules/dnn/test/test_layers.cpp#L634
these tests will cover 
```
N x C x H x W
N x C
```
https://github.com/opencv/opencv/blob/2842116ebd717356ecc8f222284161ae5f0cb4a9/modules/dnn/test/test_layers.cpp#L723
and these will cover 
```
N x H x W x C
1 x C
```

**Comment by ashishkrshrivastava** [comment]
@dkurt  do you suggest any changes in this PR. Actually changes in this PR are needed to proceed PR [#16750](https://github.com/opencv/opencv/pull/16750)

**Comment by asmorkalov** [comment]
@dkurt Do you have any comments?

File: modules/dnn/test/test_layers.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dkurt) [reply modules/dnn/test/test_layers.cpp:719]: Please keep existing test unchanged.

File: modules/dnn/src/layers/scale_layer.cpp [hunk 4]
```diff
@@ -78,64 +78,95 @@ class ScaleLayerImpl CV_FINAL : public ScaleLayer
...
         // instead of trainable weights.
-        Mat weights = hasWeights ? (blobs.empty() ? inputs[1] : blobs[0]).reshape(1, 1) : Mat();;
+        Mat weights = hasWeights ? (blobs.empty() ? inputs[1] : blobs[0]).reshape(1, 1) : Mat();
         Mat bias = hasBias ? (blobs.empty() ? inputs[1] : blobs.back()).reshape(1, 1) : Mat();
 
         MatShape inpShape = shape(inpBlob);
+        const int numSamples = inpShape[0];
+        const int numWeightsSamples = (hasWeights || hasBias) && blobs.empty() ? numSamples : 1;
         const int numWeights = !weights.empty() ? weights.total() : bias.total();
+        const int numWeightsPerSample = numWeights / numWeightsSamples;
```
Comment (dkurt) [inline modules/dnn/src/layers/scale_layer.cpp:88]: Do we have real example when weights have batch dimension? I mean when we have different NCHW input and NC weights?

**Comment by dkurt** [comment]
@ashishkrshrivastava, thanks for the patch! I still would like to recommend to write in the description what's new in this PR. It seems to me a bit overcomplicated for now and I wanted to clarify which configurations were supported before and starting from this PR.

File: modules/dnn/src/layers/scale_layer.cpp (hunk 4 above)
Comment (YashasSamaga) [reply modules/dnn/src/layers/scale_layer.cpp:88]: The scaling parameters can be computed during inference (i.e. untrainable mode). This situation will lead to different weights for different batch items.

EfficientNetB0-YOLO is one example which does it. (https://github.com/opencv/opencv/pull/16436#issuecomment-585732681)

File: modules/dnn/src/layers/scale_layer.cpp [hunk 5]
```diff
@@ -78,64 +78,95 @@ class ScaleLayerImpl CV_FINAL : public ScaleLayer
         Mat &outBlob = outputs[0];
         // There is a mode when we multiply a first blob by a second one
         // instead of trainable weights.
-        Mat weights = hasWeights ? (blobs.empty() ? inputs[1] : blobs[0]).reshape(1, 1) : Mat();;
+        Mat weights = hasWeights ? (blobs.empty() ? inputs[1] : blobs[0]).reshape(1, 1) : Mat();
         Mat bias = hasBias ? (blobs.empty() ? inputs[1] : blobs.back()).reshape(1, 1) : Mat();
 
         MatShape inpShape = shape(inpBlob);
+        const int numSamples = inpShape[0];
+        const int numWeightsSamples = (hasWeights || hasBias) && blobs.empty() ? numSamples : 1;
```
Comment (ashishkrshrivastava) [inline modules/dnn/src/layers/scale_layer.cpp:86]: Here, I assumed that if weights are non-constants, then batch size could be N for input NCHW. i.e. for each sample weight is different ( example- multiplication of two inputs ) and if weights are constants then batch size is equal to 1, assuming it will be same for all samples.

File: modules/dnn/src/layers/scale_layer.cpp [hunk 6]
```diff
@@ -78,64 +78,95 @@ class ScaleLayerImpl CV_FINAL : public ScaleLayer
...
         const int numWeights = !weights.empty() ? weights.total() : bias.total();
+        const int numWeightsPerSample = numWeights / numWeightsSamples;
         CV_Assert(numWeights != 0);
         if (hasWeights && hasBias)
             CV_CheckEQ(weights.total(), bias.total(), "Incompatible weights/bias blobs");
 
-        int endAxis;
+        axis = clamp(axis , inpBlob.dims) ? clamp(axis , inpBlob.dims) : 1;
+        CV_Assert(axis > 0 && axis < inpBlob.dims);
```
Comment (ashishkrshrivastava) [inline modules/dnn/src/layers/scale_layer.cpp:94]: if axis is 0, then it is assigned as 1. since broadcasting cannot be done at N for input dim NCHW.

File: modules/dnn/src/layers/scale_layer.cpp [hunk 7]
```diff
@@ -78,64 +78,95 @@ class ScaleLayerImpl CV_FINAL : public ScaleLayer
...
         {
-            if (total(inpShape, axis, endAxis) == numWeights)
+            if (total(inpShape, axis, endAxis) == numWeightsPerSample)
                 break;
         }
-        CV_Assert(total(inpShape, axis, endAxis) == numWeights);
+        endAxis = (numWeightsPerSample == 1) ? 1 : endAxis;
+
+        CV_Assert(total(inpShape, axis, endAxis) == numWeightsPerSample);
         CV_Assert(!hasBias || numWeights == bias.total());
```
Comment (ashishkrshrivastava) [inline modules/dnn/src/layers/scale_layer.cpp:105]: Checking broadcasting options. if size of weights is either N or 1, then endAxis=1. 
ex - NCHW ( input dims ), N/1 ( weights ). then endAxis = 1.
otherwise calculate endAxis from condition. 
`if (total(inpShape, axis, endAxis) == numWeightsPerSample)`

File: modules/dnn/src/layers/scale_layer.cpp [hunk 8]
```diff
@@ -78,64 +78,95 @@ class ScaleLayerImpl CV_FINAL : public ScaleLayer
...
         CV_CheckTypeEQ(inpBlob.type(), CV_32FC1, ""); CV_CheckTypeEQ(outBlob.type(), CV_32FC1, "");
 
-        int numSlices = total(inpShape, 0, axis);
+        int numSlices = total(inpShape, 1, axis);
         float* inpData = (float*)inpBlob.data;
```
Comment (ashishkrshrivastava) [inline modules/dnn/src/layers/scale_layer.cpp:109]: axis 0 is not considered while calculating numSlices because an extra loop for number of sample in a batch is applied at Line 116.

File: modules/dnn/src/layers/scale_layer.cpp (hunk 4 above)
Comment (ashishkrshrivastava) [reply modules/dnn/src/layers/scale_layer.cpp:88]: issue #17063 also encountered problem in EfficientNetBO ( tensorflow )

File: modules/dnn/test/test_layers.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ashishkrshrivastava) [reply modules/dnn/test/test_layers.cpp]: @dkurt Done!

[the remark this answers is not in the document]
Comment (ashishkrshrivastava) [reply modules/dnn/test/test_layers.cpp]: I didn't remove any but added some more tests.

File: modules/dnn/test/test_layers.cpp [hunk 9]
```diff
@@ -690,23 +697,94 @@ TEST_P(Scale_untrainable, Accuracy)
...
     }
     normAssert(out, ref);
 }
 
 INSTANTIATE_TEST_CASE_P(Layer_Test, Scale_untrainable, Combine(
 /*input size*/   Values(Vec4i(2, 3, 4, 5)),
 /*axis, #dims*/  Values(Vec2i(0, 1), Vec2i(0, 2), Vec2i(0, 3), Vec2i(0, 4),
+                        Vec2i(1, 1), Vec2i(1, 2), Vec2i(1, 3), Vec2i(1, 4),
+                                     Vec2i(2, 1), Vec2i(2, 2), Vec2i(2, 3),
+                                                  Vec2i(3, 1), Vec2i(3, 2)),
```
Comment (ashishkrshrivastava) [inline modules/dnn/test/test_layers.cpp:719]: These tests are for scales that are computed at runtime. So their batchsize should be equal to Batchsize of Input0. So for all of them Scales are of size [N,C,H,W] / [N,C,H] / [N,H,W] / [N,C] / [N,H] / [N,W] / [N]

Extra added tests are 
Vec2i(1,4)
Vec2i(2,3) 
Vec2i(3,2)

File: modules/dnn/test/test_layers.cpp [hunk 10]
```diff
@@ -690,23 +697,94 @@ TEST_P(Scale_untrainable, Accuracy)
...
+}
+
+INSTANTIATE_TEST_CASE_P(Layer_Test, Scale_trainable, Combine(
+/*input size*/   Values(Vec4i(2, 3, 4, 5)),
+/*axis, #dims*/  Values(Vec2i(0, 1), Vec2i(0, 2), Vec2i(0, 3),
                                      Vec2i(1, 1), Vec2i(1, 2), Vec2i(1, 3),
                                                   Vec2i(2, 1), Vec2i(2, 2),
                                                                Vec2i(3, 1)),
-/*conv fusion*/  testing::Bool()
+/*batchsize*/    testing::Bool()
```
Comment (ashishkrshrivastava) [inline modules/dnn/test/test_layers.cpp:787]: Scale trainable have weights as constants, 
Case 1: so for dim <= 4. shape would be like [1,C,H,W] / [1,C,H] / [1,H,W] / [1,C] / [1,H] / [1,W] / [1]  and 
Case 2: for dim<=3, shape would be [C,H,W] / [C,H] / [H,W] / [C] / [H] / [W].
One more argument `batchsize` is added.
batchsize = True for Case 1
batchsize = False for Case 2

-------------------------------------------Batchsize=True---------------------------------------Batchsize=False------------------------------
Vec2i(0,1)-----------------------------------[1,C]-----------------------------------------------------[C]-------------------------------------------
Vec2i(0,2)-----------------------------------[1,C,H]--------------------------------------------------[C,H]----------------------------------------
Vec2i(0,3)-----------------------------------[1,C,H,W]----------------------------------------------[C,H,W]------------------------------------
**XX** Vec2i(0,4)------------------------------[1,C,H,W,_]----------------------------------------------[C,H,W,_]----------------------------
Vec2i(1,1)-----------------------------------[1,C]-----------------------------------------------------[C]-------------------------------------------
Vec2i(1,2)-----------------------------------[1,C,H]--------------------------------------------------[C,H]----------------------------------------
Vec2i(1,3)-----------------------------------[1,C,H,W]----------------------------------------------[C,H,W]-------------------------------------
Vec2i(2,1)-----------------------------------[1,H]-----------------------------------------------------[H]-------------------------------------------
Vec2i(2,2)-----------------------------------[1,C,H]--------------------------------------------------[C,H]-----------------------------------------
Vec2i(3,1)-----------------------------------[1,H]-----------------------------------------------------[H]--------------------------------------------

So all of the cases are there but Vec2i(0,4)

**Comment by asmorkalov** [comment]
@ashishkrshrivastava @dkurt Do you have conclusion on the patch? Is it ready?

**Comment by dkurt** [comment]
Yeah, sorry for delay. Just wanted to make some local experiments.

**Comment by dkurt** [comment]
@ashishkrshrivastava, thanks for the complete solution for these three issues! It's good to see that it works but I wanted to propose smaller some steps to solve them one-by-one and probably it will help us to simplify this PR and additionally fix other bugs/weak points.

1. issue https://github.com/opencv/opencv/issues/13626 can be solved by a single line patch:
```patch
@@ -303,7 +303,7 @@ Ptr<Layer> ShiftLayer::create(const LayerParams& params)
     scaleParams.type = "Scale";
     scaleParams.blobs = params.blobs;
     scaleParams.set("bias_term", true);
-    scaleParams.set("axis", 0);
+    scaleParams.set("axis", 1);
```

<details>
<summary>validation code</summary>

```python
import numpy as np
import cv2 as cv

inp0 = cv.dnn.blobFromImage(cv.imread('/home/dkurt/Pictures/image.jpg'), size=(300, 300), mean=(104, 177, 123))
inp1 = cv.dnn.blobFromImage(cv.imread('/home/dkurt/Pictures/image2.jpg'), size=(300, 300), mean=(104, 177, 123))

batch = np.concatenate([inp0, inp1])

net = cv.dnn.readNet('opencv_face_detector_uint8.pb', 'opencv_face_detector.pbtxt')
net.setInput(batch)
out = net.forward()
print(out.shape)

net = cv.dnn.readNet('opencv_face_detector_uint8.pb', 'opencv_face_detector.pbtxt')
net.setInput(inp0)
out0 = net.forward()
net.setInput(inp1)
out1 = net.forward()

out = out.reshape(-1, 7)[:,1:]  # remove batch id column
out0 = out0.reshape(-1, 7)[:,1:]
out1 = out1.reshape(-1, 7)[:,1:]

num0 = 0
for det in out0:
    if det[0] != 0:  # exclude zeroed predictions at the bottom
        num0 += 1
    else:
        break
print(np.max(np.abs(out[:num0] - out0[:num0])))

num1 = 0
for det in out1:
    if det[0] != 0:  # exclude zeroed predictions at the bottom
        num1 += 1
    else:
        break
print(np.max(np.abs(out[num0:(num0+num1)] - out1[:num1])))
```

</details>

The only problem is failed test `runTensorFlowNet("shift_reshape_no_reorder");`:
```python
inp = tf.placeholder(tf.float32, [2, 3, 4], 'input')
shift = tf.Variable(tf.random_normal([2, 3, 4]), name='shift')
shifted = tf.add(inp, shift, name='shifted')
reshape = tf.reshape(shifted, [4, 3, 2], 'reshaped')
save(inp, reshape, 'shift_reshape_no_reorder')
```

**Comment by asmorkalov** [comment]
@dkurt @ashishkrshrivastava Have you driven to conclusion with the patch?

**Comment by ashishkrshrivastava** [comment]
> 1. issue #13626 can be solved by a single line patch:
> 
> ```diff
> @@ -303,7 +303,7 @@ Ptr<Layer> ShiftLayer::create(const LayerParams& params)
>      scaleParams.type = "Scale";
>      scaleParams.blobs = params.blobs;
>      scaleParams.set("bias_term", true);
> -    scaleParams.set("axis", 0);
> +    scaleParams.set("axis", 1);
> ```
> 
> validation code
> The only problem is failed test `runTensorFlowNet("shift_reshape_no_reorder");`:

As you change axis default to 1.
Broadcasting checking will change 
```cpp
1. for (endAxis = axis + 1; endAxis <= inpBlob.dims; ++endAxis)
2.        {
3.            if (total(inpShape, axis, endAxis) == numWeights)
4.                break;
5.        }
```
at **line 3**
I guess, that is why it is failing.

> but I wanted to propose some smaller steps to solve them one-by-one and probably it will help us to simplify this PR and additionally fix other bugs/weak points.

@dkurt you are right that the changes are complex. I have given the explanations above as reviews. 
I think all the issues has same problem. I have checked how to solve them one-by-one but I think the solution cant be breakdown into parts.

**Comment by asmorkalov** [comment]
@dkurt Friendly reminder.

**Comment by asmorkalov** [comment]
@dkurt Reminder.

**Comment by asmorkalov** [comment]
@dkurt Reminder.

**Comment by asmorkalov** [comment]
@dkurt, @ashishkrshrivastava  Do you have conclusion on the patch?

**Comment by liuxingbaoyu** [comment]
Hello, are there any updates?
@dkurt @alalek Sorry to interrupt.

**Comment by asmorkalov** [comment]
@dkurt @ashishkrshrivastava Friendly reminder.

**Comment by asmorkalov** [comment]
@dkurt Friendly reminder.

**Comment by asmorkalov** [comment]
@dkurt Friendly reminder.

**Comment by ashishkrshrivastava** [comment]
@dkurt @asmorkalov are there any improvements needed or should I close this PR so that if anyone wants to work on this issue can contribute?

**Review by alalek — COMMENTED** [review]
@ashishkrshrivastava Thank you for the contribution!

Could you please add some link on used specification (reference code) for this layer?
Part with axis 0=>1 should is not clear to me yet.

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by ashishkrshrivastava** [comment]
@alalek , without changing tests these modifications won't work. I found that both tests and implementation were incorrect so I modified both. Maybe the way I was approaching this issue was wrong,  but these modifications were solving all issues related to this. I am closing this PR so that anyone else can work on it.
